### PR TITLE
Put all debug/developer behind an AlmostFinalValue false

### PR DESCRIPTION
**Pull request by dwijnand**
Builds on #9571 and #9596.

### CODE REVIEW

**Comment by dwijnand** [comment]
https://scala-ci.typesafe.com/job/compiler-benchmark/3673/console

```
[info] Benchmark                                                (scalaVersion)  (source)    Score   Error  Units
[info] HotScalacBenchmark.compile                  2.13.6-bin-cd03dfd-SNAPSHOT    scalap  627.856 ± 1.295  ms/op
[info] HotScalacBenchmark.compile:compile·p0.00    2.13.6-bin-cd03dfd-SNAPSHOT    scalap  614.466          ms/op
[info] HotScalacBenchmark.compile:compile·p0.50    2.13.6-bin-cd03dfd-SNAPSHOT    scalap  624.951          ms/op
[info] HotScalacBenchmark.compile:compile·p0.90    2.13.6-bin-cd03dfd-SNAPSHOT    scalap  640.680          ms/op
[info] HotScalacBenchmark.compile:compile·p0.95    2.13.6-bin-cd03dfd-SNAPSHOT    scalap  648.020          ms/op
[info] HotScalacBenchmark.compile:compile·p0.99    2.13.6-bin-cd03dfd-SNAPSHOT    scalap  657.625          ms/op
[info] HotScalacBenchmark.compile:compile·p0.999   2.13.6-bin-cd03dfd-SNAPSHOT    scalap  666.894          ms/op
[info] HotScalacBenchmark.compile:compile·p0.9999  2.13.6-bin-cd03dfd-SNAPSHOT    scalap  666.894          ms/op
[info] HotScalacBenchmark.compile:compile·p1.00    2.13.6-bin-cd03dfd-SNAPSHOT    scalap  666.894          ms/op
```

```
[info] Benchmark                                                (scalaVersion)  (source)    Score   Error  Units
[info] HotScalacBenchmark.compile                  2.13.6-bin-cd03dfd-SNAPSHOT    scalap  680.978 ± 1.439  ms/op
[info] HotScalacBenchmark.compile:compile·p0.00    2.13.6-bin-cd03dfd-SNAPSHOT    scalap  658.506          ms/op
[info] HotScalacBenchmark.compile:compile·p0.50    2.13.6-bin-cd03dfd-SNAPSHOT    scalap  679.477          ms/op
[info] HotScalacBenchmark.compile:compile·p0.90    2.13.6-bin-cd03dfd-SNAPSHOT    scalap  693.109          ms/op
[info] HotScalacBenchmark.compile:compile·p0.95    2.13.6-bin-cd03dfd-SNAPSHOT    scalap  698.352          ms/op
[info] HotScalacBenchmark.compile:compile·p0.99    2.13.6-bin-cd03dfd-SNAPSHOT    scalap  707.254          ms/op
[info] HotScalacBenchmark.compile:compile·p0.999   2.13.6-bin-cd03dfd-SNAPSHOT    scalap  715.129          ms/op
[info] HotScalacBenchmark.compile:compile·p0.9999  2.13.6-bin-cd03dfd-SNAPSHOT    scalap  715.129          ms/op
[info] HotScalacBenchmark.compile:compile·p1.00    2.13.6-bin-cd03dfd-SNAPSHOT    scalap  715.129          ms/op
```

https://scala-ci.typesafe.com/job/compiler-benchmark/3672/console

```
[info] Benchmark                                                (scalaVersion)  (source)    Score   Error  Units
[info] HotScalacBenchmark.compile                  2.13.6-bin-50f08c6-SNAPSHOT    scalap  626.077 ± 1.576  ms/op
[info] HotScalacBenchmark.compile:compile·p0.00    2.13.6-bin-50f08c6-SNAPSHOT    scalap  607.126          ms/op
[info] HotScalacBenchmark.compile:compile·p0.50    2.13.6-bin-50f08c6-SNAPSHOT    scalap  623.903          ms/op
[info] HotScalacBenchmark.compile:compile·p0.90    2.13.6-bin-50f08c6-SNAPSHOT    scalap  639.631          ms/op
[info] HotScalacBenchmark.compile:compile·p0.95    2.13.6-bin-50f08c6-SNAPSHOT    scalap  645.608          ms/op
[info] HotScalacBenchmark.compile:compile·p0.99    2.13.6-bin-50f08c6-SNAPSHOT    scalap  662.847          ms/op
[info] HotScalacBenchmark.compile:compile·p0.999   2.13.6-bin-50f08c6-SNAPSHOT    scalap  668.991          ms/op
[info] HotScalacBenchmark.compile:compile·p0.9999  2.13.6-bin-50f08c6-SNAPSHOT    scalap  668.991          ms/op
[info] HotScalacBenchmark.compile:compile·p1.00    2.13.6-bin-50f08c6-SNAPSHOT    scalap  668.991          ms/op
```

```
[info] Benchmark                                                (scalaVersion)  (source)    Score   Error  Units
[info] HotScalacBenchmark.compile                  2.13.6-bin-50f08c6-SNAPSHOT    scalap  689.112 ± 1.765  ms/op
[info] HotScalacBenchmark.compile:compile·p0.00    2.13.6-bin-50f08c6-SNAPSHOT    scalap  663.749          ms/op
[info] HotScalacBenchmark.compile:compile·p0.50    2.13.6-bin-50f08c6-SNAPSHOT    scalap  688.914          ms/op
[info] HotScalacBenchmark.compile:compile·p0.90    2.13.6-bin-50f08c6-SNAPSHOT    scalap  703.594          ms/op
[info] HotScalacBenchmark.compile:compile·p0.95    2.13.6-bin-50f08c6-SNAPSHOT    scalap  708.837          ms/op
[info] HotScalacBenchmark.compile:compile·p0.99    2.13.6-bin-50f08c6-SNAPSHOT    scalap  721.420          ms/op
[info] HotScalacBenchmark.compile:compile·p0.999   2.13.6-bin-50f08c6-SNAPSHOT    scalap  728.760          ms/op
[info] HotScalacBenchmark.compile:compile·p0.9999  2.13.6-bin-50f08c6-SNAPSHOT    scalap  728.760          ms/op
[info] HotScalacBenchmark.compile:compile·p1.00    2.13.6-bin-50f08c6-SNAPSHOT    scalap  728.760          ms/op
```

**Comment by dwijnand** [comment]
Using https://github.com/scala/scala/pull/9564#issuecomment-818743175 for reference and target values, for JDK 8 (the first of the two code blocks), the reference value is 628.529 ± 1.312 ms/op, the target is 627.308 ± 1.274, we got 627.856 ± 1.295 and 626.077 ± 1.576, which are all too close to the margin of error to be callable.

For JDK 11, the reference is 690.520 ± 1.643 ms/op, the target is 680.190 ± 1.242 and we hit 680.978 ± 1.439 and 689.112 ± 1.765, which means the first commit hits our target, while the second one maintains a 9 ms/op of overhead (out of the 10 we were trying to lose) 😞

File: src/reflect/scala/reflect/internal/settings/MutableSettings.scala [hunk 1]
```diff
@@ -66,6 +68,9 @@ abstract class MutableSettings extends AbsSettings {
 
   def isScala212: Boolean
   private[scala] def isScala213: Boolean
+
+  @inline final def isDebug: Boolean     = StatisticsStatics.isDebug     && debug
```
Comment (viktorklang) [inline src/reflect/scala/reflect/internal/settings/MutableSettings.scala]: @dwijnand Which machinery enables treating BooleanSettings as `boolean`s? (and what is the performance impact of that?)

Comment (dwijnand) [reply src/reflect/scala/reflect/internal/settings/MutableSettings.scala]:
```scala
object MutableSettings {
  import scala.language.implicitConversions
  /** Support the common use case, `if (settings.debug) println("Hello, martin.")` */
  @inline implicit def reflectSettingToBoolean(s: MutableSettings#BooleanSetting): Boolean = s.value
}
```

Comment (viktorklang) [reply src/reflect/scala/reflect/internal/settings/MutableSettings.scala]: And `value` is a non-final, non-@inline method which then calls a non-final `var`s non-@inline accessor method?
Seems like a lot of indirection to get a true or a false from an instance? Does it make any difference performance-wise if you improve on that access-pattern?

**Comment by lrytz** [comment]
The inliner is doing its job (on the last commit). There are some differences in the bytecode, but it's all things that I'd hope the JVM could easily handle...

Here's the bytecode for
```
  @inline final override def debuglog(msg: => String): Unit = {
    if (StatisticsStatics.isDebug && settings.debug)
      log(msg)
  }
```

```
  public final debuglog(Lscala/Function0;)V
    // parameter final  msg
   L0
    LINENUMBER 317 L0
    INVOKESTATIC scala/reflect/internal/util/StatisticsStatics.isDebug ()Z
    IFEQ L1
    GETSTATIC scala/reflect/internal/settings/MutableSettings$.MODULE$ : Lscala/reflect/internal/settings/MutableSettings$;
    POP
    ALOAD 0
    INVOKEVIRTUAL scala/tools/nsc/Global.settings ()Lscala/tools/nsc/Settings;
    INVOKEVIRTUAL scala/tools/nsc/Settings.debug ()Lscala/tools/nsc/settings/MutableSettings$BooleanSetting;
    INVOKEINTERFACE scala/reflect/internal/settings/MutableSettings$SettingValue.value ()Ljava/lang/Object; (itf)
    INVOKESTATIC scala/runtime/BoxesRunTime.unboxToBoolean (Ljava/lang/Object;)Z
    IFEQ L1
...
```

and with the inlined `if (settings.isDebug)` 

```
  public final debuglog(Lscala/Function0;)V
    // parameter final  msg
   L0
    LINENUMBER 316 L0
    ALOAD 0
    INVOKEVIRTUAL scala/tools/nsc/Global.settings ()Lscala/tools/nsc/Settings;
    DUP
    IFNONNULL L1
    ACONST_NULL
    ATHROW
   L1
   FRAME SAME1 scala/tools/nsc/Settings
    ASTORE 2
   L2
    INVOKESTATIC scala/reflect/internal/util/StatisticsStatics.isDebug ()Z
    IFEQ L3
    GETSTATIC scala/reflect/internal/settings/MutableSettings$.MODULE$ : Lscala/reflect/internal/settings/MutableSettings$;
    POP
    ALOAD 2
    INVOKEVIRTUAL scala/reflect/internal/settings/MutableSettings.debug ()Lscala/reflect/internal/settings/MutableSettings$SettingValue;
    INVOKEINTERFACE scala/reflect/internal/settings/MutableSettings$SettingValue.value ()Ljava/lang/Object; (itf)
    INVOKESTATIC scala/runtime/BoxesRunTime.unboxToBoolean (Ljava/lang/Object;)Z
    IFEQ L3
    ICONST_1
    ISTORE 3
    GOTO L4
   L3
   FRAME APPEND [scala/tools/nsc/Settings]
    ICONST_0
    ISTORE 3
   L4
   FRAME APPEND [I]
    ILOAD 3
    ACONST_NULL
    ASTORE 2
    IFEQ L5
...
```

Differences I can see:
  - `null` check on `this.settings`
  - `this.settings` is stored into a local, the local is set to `null` at the end
  - the result (boolean) is stored into a local, some additional `goto`s around that

**Comment by som-snytt** [comment]
Is the upshot that "use BooleanSetting as a Boolean" will always be too slow, so special-case debug to make it more free?

Along similar lines, I remember noticing how expensive `sym != NoSymbol` is, where `sym ne NoSymbol` is not an optimization, because all the work is getting `NoSymbol` out of the cake. That makes it a galette des rois, where the kids go fishing for the prize.

**Comment by dwijnand** [comment]
Too slow when used for things that are used many times (stats, debug) but overwhelmingly mostly false.

**Comment by retronym** [comment]
Rebased to resolve a merge conflict.

File: src/compiler/scala/tools/nsc/backend/jvm/PostProcessorFrontendAccess.scala [hunk 2]
```diff
@@ -184,7 +184,7 @@ object PostProcessorFrontendAccess {
     private def buildCompilerSettings(): CompilerSettings = new CompilerSettings {
       import global.{settings => s}
 
-      val debug: Boolean = s.debug
+      @inline final val debug: Boolean = s.isDebug
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/backend/jvm/PostProcessorFrontendAccess.scala]: This should be an `@inline def`.

**Comment by som-snytt** [comment]
Not sure where to say yay.

**Pull request merged**